### PR TITLE
GMP Documentation, Alerts, and Dashboard filters: Apache Flink

### PR DESCRIPTION
**Pull request by algchoo**
This PR includes the following for the Flink GMP integration:
* documentation
* alerts
* dashboard filters

### CODE REVIEW

File: integrations/flink/prometheus_metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,212 @@
+platforms:
```
Comment (varun-c) [inline integrations/flink/prometheus_metadata.yaml:1]: The flink-task-manager-prometheus dashboard (https://github.com/GoogleCloudPlatform/monitoring-dashboard-samples/blob/2de3d85fcd11d86dc3d7429075287ac189fbb88c/dashboards/flink/flink-task-manager-prometheus.json) uses the following metrics that are missing in here:

- flink_taskmanager_Status_JVM_GarbageCollector_G1_Old_Generation_Time
- flink_taskmanager_Status_JVM_GarbageCollector_G1_Young_Generation_Time
- flink_taskmanager_Status_JVM_GarbageCollector_G1_Old_Generation_Count
- flink_taskmanager_Status_JVM_GarbageCollector_G1_Young_Generation_Count

Should these metrics be included here?

**Review by varun-c — COMMENTED** [review]
Reviewed the flink-job-manager-prometheus.json, flink-task-manager-prometheus.json and prometheus_metadata.yaml files. They look good to me but I have one comment about some missing metrics.

Adding yqlu to review the documentation.yaml.

File: integrations/flink/documentation.yaml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+app_site_name: Flink
+app_site_url: https://flink.apache.org/
+exporter_name: the Flink exporter
+exporter_pkg_name: Flink
+exporter_repo_url: https://flink.apache.org/features/2019/03/11/prometheus-monitoring.html
+additional_prereq_info: |
+  {{app_name_short}} exposes Prometheus-format metrics when configured with
+  `metrics.reporter.prom.class: org.apache.flink.metrics.prometheus.PrometheusReporter`.
+
+   *  To determine if {{app_name_short}} is exposing metrics, run the following command:
```
Comment (yqlu) [inline integrations/flink/documentation.yaml:13]: Actually, I could not get metrics to show up by using the [sample getting started manifests](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/resource-providers/standalone/kubernetes/)... I got:

```
2023-01-09 21:26:20,313 ERROR org.apache.flink.runtime.metrics.ReporterSetup               [] - Could not instantiate metrics reporter prom. Metrics might not be exposed/reported.
java.lang.ClassNotFoundException: org.apache.flink.metrics.prometheus.PrometheusReporter
        at jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source) ~[?:?]
        at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source) ~[?:?]
        at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:?]
        at java.lang.Class.forName0(Native Method) ~[?:?]
        at java.lang.Class.forName(Unknown Source) ~[?:?]
        at org.apache.flink.runtime.metrics.ReporterSetup.loadViaReflection(ReporterSetup.java:456) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.metrics.ReporterSetup.loadReporter(ReporterSetup.java:409) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.metrics.ReporterSetup.setupReporters(ReporterSetup.java:328) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.metrics.ReporterSetup.fromConfiguration(ReporterSetup.java:209) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.entrypoint.ClusterEntrypoint.createMetricRegistry(ClusterEntrypoint.java:458) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.entrypoint.ClusterEntrypoint.initializeServices(ClusterEntrypoint.java:397) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.entrypoint.ClusterEntrypoint.runCluster(ClusterEntrypoint.java:282) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.entrypoint.ClusterEntrypoint.lambda$startCluster$1(ClusterEntrypoint.java:232) ~[flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.security.contexts.NoOpSecurityContext.runSecured(NoOpSecurityContext.java:28) [flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.entrypoint.ClusterEntrypoint.startCluster(ClusterEntrypoint.java:229) [flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.entrypoint.ClusterEntrypoint.runClusterEntrypoint(ClusterEntrypoint.java:729) [flink-dist-1.16.0.jar:1.16.0]
        at org.apache.flink.runtime.entrypoint.StandaloneSessionClusterEntrypoint.main(StandaloneSessionClusterEntrypoint.java:59) [flink-dist-1.16.0.jar:1.16.0]
2023-01-09 21:26:20,328 INFO  org.apache.flink.runtime.metrics.MetricRegistryImpl          [] - No metrics reporter configured, no metrics will be exposed/reported.
```

Weirdly, after referencing the OiQ yaml manifests and inspecting the diff, I got this to work by changing
```
      - name: jobmanager
        # image: apache/flink:latest
        image: apache/flink:1.15.2-scala_2.12 # fixed
        args: ["jobmanager"]
```

Did you experience the same? Also, I noticed that the OiQ yaml manifests have
```
        ports:
        ...
        - containerPort: 9249
          name: prometheus     
```

Is this necessary?

File: integrations/flink/documentation.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+      <pre class="devsite-click-to-copy">
+      kubectl -n <namespace_name> port-forward <flink_pod_name> 9249
+      curl localhost:9249/metrics
+      </pre>
+dashboard_available: true
+multiple_dashboards: false
```
Comment (yqlu) [inline integrations/flink/documentation.yaml:20]: multiple_dashboards should be true?

File: integrations/flink/prometheus_metadata.yaml [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+  - type: GKE
+    launch_stage: GA
```
Comment (yqlu) [inline integrations/flink/prometheus_metadata.yaml:3]: Hidden

File: integrations/flink/documentation.yaml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+additional_prereq_info: |
+  {{app_name_short}} exposes Prometheus-format metrics when configured with
+  `metrics.reporter.prom.class: org.apache.flink.metrics.prometheus.PrometheusReporter`.
+
+   *  To determine if {{app_name_short}} is exposing metrics, run the following command:
+
+      <pre class="devsite-click-to-copy">
+      kubectl -n <namespace_name> port-forward <flink_pod_name> 9249
+      curl localhost:9249/metrics
+      </pre>
```
[the remark this answers is not in the document]
Comment (mtabasko) [reply integrations/flink/documentation.yaml:18]: Suggest using my suggestion from the comment on the Velero PR (ideally these steps should be numbered, which the RabbitMQ YAML referenced here does not do; I added numbering manually when I noticed it.)

File: integrations/flink/documentation.yaml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    endpoints:
+  + - port: 9249
+      scheme: http
+      interval: 30s
+      path: /metrics
+    selector:
+      matchLabels:
+  +     app.kubernetes.io/name: flink
+additional_podmonitoring_info: |
+  Ensure that the port and matchLabels match that of the {{app_name_short}} pods you wish to monitor.
```
Comment (mtabasko) [inline integrations/flink/documentation.yaml:40]: Ensure that the values of the \`port\` and \`matchLabels\` fields match thse of the {{app_name_short}} pods you want to monitor.

**Review by mtabasko — CHANGES_REQUESTED** [review]

File: integrations/flink/documentation.yaml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+   *  To determine if {{app_name_short}} is exposing metrics, run the following command:
+
+      <pre class="devsite-click-to-copy">
+      kubectl -n <namespace_name> port-forward <flink_pod_name> 9249
+      curl localhost:9249/metrics
+      </pre>
+dashboard_available: true
+multiple_dashboards: false
+dashboard_display_name: {{app_name_short}} Prometheus Overview
```
[the remark this answers is not in the document]
Comment (mtabasko) [reply integrations/flink/documentation.yaml:21]: Looks like 1.15, based on "prometheus_metadata.yaml", line 10.

File: integrations/flink/documentation.yaml (hunk 2 above)
Comment (jsirianni) [reply integrations/flink/documentation.yaml:13]: I would expect "latest" to work, however, I am not a fan of using latest outside of a quick test environment.

With that said, I am getting metrics with our dev configs, without changes. If I remove the `prometheus` named port from the pod spec's `ports`, I can still access the port w/ port forwarding. I just loose the ability to reference the port by name, which we are not doing anywhere.

File: integrations/flink/documentation.yaml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  apiVersion: monitoring.googleapis.com/v1
+  kind: PodMonitoring
+  metadata:
+    name: flink
+    labels:
+      app.kubernetes.io/name: flink
+      app.kubernetes.io/part-of: google-cloud-managed-prometheus
+  spec:
+    endpoints:
+  + - port: 9249
```
[the remark this answers is not in the document]
Comment (jsirianni) [reply integrations/flink/documentation.yaml:32]: Correct. The user may or may not have this port defined in their pod spec, but the Prometheus scraper can still access it as long as Prometheus metrics are enabled.

File: integrations/flink/documentation.yaml [hunk 9]
```diff
@@ -4,20 +4,61 @@ app_name: {{app_name_short}}
...
+    name: basic-example
+  spec:
+    image: flink:1.15
+    flinkVersion: v1_15
+    flinkConfiguration:
+      taskmanager.numberOfTaskSlots: "2"
+  +   metrics.reporter.prom.class: org.apache.flink.metrics.prometheus.PrometheusReporter
+  </pre>
+
+  Alternatively, you can specify the prometheus reporter as a default option within the [Flink operator
```
Comment (yqlu) [inline integrations/flink/documentation.yaml:46]: prometheus --> Prometheus (general rule of thumb?)

File: integrations/flink/documentation.yaml (hunk 2 above)
Comment (yqlu) [reply integrations/flink/documentation.yaml]: OK, sounds good.

File: integrations/flink/documentation.yaml (hunk 3 above)
Comment (yqlu) [reply integrations/flink/documentation.yaml]: Tagging @mtabasko  -- we don't have many docs with `multiple_dashboards = true` -- in this case, the template no longer uses `dashboard_display_name`, right? If so, let's clean it up.

File: integrations/flink/documentation.yaml [hunk 10]
```diff
@@ -0,0 +1,79 @@
+exporter_type: included
+app_name_short: Flink
+app_name: {{app_name_short}}
```
[the remark this answers is not in the document]
Comment (yqlu) [reply integrations/flink/documentation.yaml:3]: Ah sorry, just realized this was marked resolved but not yet changed.

[the remark this answers is not in the document]
Comment (yqlu) [reply integrations/flink/documentation.yaml:3]: Let's be consistent with what was done in #372 -- such that app_name_short is still Flink, and then app_name is Apache {{app_name_short}}

[the remark this answers is not in the document]
Comment (algchoo) [reply integrations/flink/documentation.yaml]: Understood, this should be good to go across all Apache related PRs

**Pull request merged**